examples/cpp/pyperf/PyPerfBPFProgram.cc

```diff
@@ -313,6 +317,43 @@ on_event(struct pt_regs* ctx) {
   event->stack_status = STACK_STATUS_ERROR;
   event->error_code = ERROR_NONE;
 
+  struct task_struct const *const task = (struct task_struct *)bpf_get_current_task();
+
+  // Get raw native user stack
+  struct pt_regs user_regs;
+
+  // ebpf doesn't allow direct access to ctx->cs, so we need to copy it
+  int cs;
+  bpf_probe_read_kernel(&cs, sizeof(cs), &(ctx->cs));
+
+  // Are we in user mode?
+  if (cs & 3) {
+    user_regs = *ctx;
+  }
+  else {
+    // The third argument is equivalent to `task_pt_regs(task)` for x86. Macros doesn't
+    // work properly on bcc, so we need to re-implement.
+    bpf_probe_read_kernel(
+        &user_regs, sizeof(user_regs),
+        (struct pt_regs *)((unsigned long)(task->stack) + THREAD_SIZE -
+                           TOP_OF_KERNEL_STACK_PADDING) - 1);
+  }
+
+  event->user_sp = user_regs.sp;
+  event->user_ip = user_regs.ip;
+
+  // Copy one page at the time - if one fails we don't want to lose the others
+  int i;
+  #pragma unroll
+  for (i = 0; i < __USER_STACKS_PAGES__; ++i) {
+    if (bpf_probe_read_user(
+            event->raw_user_stack + i * PAGE_SIZE, PAGE_SIZE,
+            (void *)((user_regs.sp & PAGE_MASK) + (i * PAGE_SIZE))) < 0) {
+        break;
+      }
+  }
+  event->user_stack_len = i * PAGE_SIZE;
```

Comment:
Stick this in the loop and the rhs will be entirely constified.

Reply:
I'm not sure what do you mean or what is it good for

Reply:
I mean if you place this statement inside the loop at the beginning:

```c
   for (i = 0; i < __USER_STACKS_PAGES__; ++i) {
       event->user_stack_len = i * PAGE_SIZE;
```

Then after unrolling `i` will be replaced with the iteration number (`0` through `__USER_STACKS_PAGES__`) and the entire right-hand side expression will be evaluated at compile-time. So you would get bytecode that roughly does:

```
event->user_stack_len = 0;
...
event->user_stack_len = 4096;
...
event->user_stack_len = 8192;
...
```

Instead of:

```
i = 0;
...
i = 1;
...
i = 2;
...
...
event->user_stack_len = i * 4096;
```